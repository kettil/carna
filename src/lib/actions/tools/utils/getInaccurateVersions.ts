import { isObject, objectEntries } from '@kettil/tool-lib';
import { coerce } from 'semver';

const regexpPostfix = /\d+\.\d+\.\d+[-.][\d\-.a-z]+/iu;

const getInaccurateVersions = (dependencies: unknown): Array<[string, string, string]> => {
  if (!isObject(dependencies)) {
    return [];
  }

  return objectEntries(dependencies)
    .filter((value): value is readonly [string, string] => typeof value[1] === 'string')
    .filter(([, version]) => !version.startsWith('file'))
    .filter(([, version]) => !version.startsWith('http'))
    .filter(([, version]) => !version.startsWith('git'))
    .map<[string, string, string]>(([name, version]) => [
      name,
      version,
      coerce(version)?.version ?? 'could not be detected',
    ])
    .filter(([, version, result]) => version !== result)
    .filter(([, version]) => !regexpPostfix.test(version));
};

export { getInaccurateVersions };
