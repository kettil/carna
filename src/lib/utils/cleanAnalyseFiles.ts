import { uniqueArray } from '@kettil/tool-lib';
import { eslintExtensions, prettierExtensions } from '../../configs/actionConfigs';

const testEslint = new RegExp(`(${eslintExtensions.replaceAll(',', '|')})$`, 'u');
const testPrettier = new RegExp(`(${prettierExtensions.replaceAll(',', '|')})$`, 'u');

const cleanAnalyseFiles = (
  files: string[],
): { eslintFiles: string[]; prettierFiles: string[]; mergedFiles: string[] } => {
  const eslintFiles = files.filter((file) => testEslint.test(file));
  const prettierFiles = files.filter((file) => testPrettier.test(file));
  const mergedFiles = uniqueArray([...prettierFiles, ...eslintFiles]);

  return { eslintFiles, prettierFiles, mergedFiles };
};

export { cleanAnalyseFiles };
