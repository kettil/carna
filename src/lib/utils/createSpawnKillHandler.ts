import { constants } from 'os';
import { getStdin } from '../cli/process';

type SpawnKillCallback = (signal: NodeJS.Signals | number | undefined) => boolean;

type SpawnKillHandler = {
  addCallback: (callback: SpawnKillCallback) => void;
  kill: (signal: NodeJS.Signals | number | undefined) => boolean[];
};

const createSpawnKillHandler = ({ registerStdin = false }: { registerStdin?: boolean } = {}): SpawnKillHandler => {
  const callbacks: SpawnKillCallback[] = [];
  let isRunning = true;

  if (registerStdin) {
    const stdin = getStdin();

    stdin.setRawMode(true);
    stdin.setEncoding('utf8');
    stdin.resume();

    stdin.on('data', (key) => {
      // ctrl-c ( end of text )
      if (key.toString() === '\u0003') {
        stdin.setRawMode(false);
        stdin.pause();

        callbacks.forEach((callback) => callback(constants.signals.SIGINT));
      }
    });
  }

  return {
    addCallback: (callback) => callbacks.push(callback),
    kill: (signal) => {
      if (isRunning) {
        isRunning = false;

        return callbacks.map((callback) => callback(signal));
      }

      return [];
    },
  };
};

export type { SpawnKillHandler };
export { createSpawnKillHandler };
