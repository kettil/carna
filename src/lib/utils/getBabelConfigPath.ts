import { babelConfigFiles } from '../../configs/actionConfigs';
import { FirstExistFileError } from '../errors/firstExistFileError';
import { getFirstExistingFile } from './getFirstExistingFile';

const getBabelConfigPath = async ([cwd, ...paths]: string[]): Promise<string> => {
  try {
    return await getFirstExistingFile({ cwd, files: babelConfigFiles });
  } catch (error) {
    if (error instanceof FirstExistFileError && paths.length > 0) {
      return getBabelConfigPath(paths);
    }

    throw error;
  }
};

export { getBabelConfigPath };
