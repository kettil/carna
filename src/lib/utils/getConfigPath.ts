import { babelConfigFiles } from '../../configs/actionConfigs';
import { getFirstExistingFile } from './getFirstExistingFile';

const getBabelConfigPath = async ({ cwd }: { cwd: string }): Promise<string> =>
  getFirstExistingFile({ cwd, files: babelConfigFiles });

export { getBabelConfigPath };
