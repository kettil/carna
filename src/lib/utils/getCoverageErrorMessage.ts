import { bold, red } from 'chalk';
import type { CoverageSummary } from 'istanbul-lib-coverage';
import type { Watermark, Watermarks } from 'istanbul-lib-report';
import { getCoverageWatermark } from './getCoverageWatermark';

const getCoverageErrorMessage = (
  key: keyof Watermarks,
  watermarks: Partial<Record<keyof Watermarks, Watermark | number>>,
  coverageSummary: CoverageSummary,
): string => {
  const limit = Math.min(...getCoverageWatermark(watermarks[key]));
  const value = coverageSummary[key].pct;

  return red(`Coverage threshold for ${bold(key)} (${limit}%) not met: ${value}%`);
};

export { getCoverageErrorMessage };
