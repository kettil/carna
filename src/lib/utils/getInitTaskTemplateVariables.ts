import { TemplateVariable } from '../actions/node/template';
import { InitSettings } from './getInitTaskSettings';

const getInitTaskTemplateVariables = (settings: Pick<InitSettings, 'github'>): Record<string, TemplateVariable> => {
  /* eslint-disable @typescript-eslint/naming-convention */
  const variables: Record<string, TemplateVariable> = {
    GITHUB_USERNAME: settings.github.name,
  };

  return variables;
};

export { getInitTaskTemplateVariables };
