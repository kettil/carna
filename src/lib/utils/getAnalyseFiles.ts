import { join } from 'path';
import { promisify } from 'util';
import { isString, uniqueArray } from '@kettil/tool-lib';
import { glob } from 'glob';
import { gitLsAction } from '../actions/git/ls';
import { gitStagedAction } from '../actions/git/staged';
import { access } from '../cmd/access';
import { existFiles } from '../cmd/existFiles';
import type { Task } from '../types';
import { cleanAnalyseFiles } from './cleanAnalyseFiles';

const globPromise = promisify(glob);

type AnalyseFiles = {
  prettierFiles: string[];
  eslintFiles: string[];
};

const getAnalyseFiles: Task<{ path?: string; all?: boolean; files?: AnalyseFiles }, AnalyseFiles | undefined> = async (
  argv,
  { files, all, path },
) => {
  if (all || files) {
    return files;
  }

  if (isString(path)) {
    // get all files from "path"
    const pathFiles = await globPromise(`${path}/**/*`, { cwd: argv.cwd, absolute: true });

    return cleanAnalyseFiles(pathFiles);
  }

  const hasGitFolder = await access(join(argv.root, '.git'), 'readable');

  if (!hasGitFolder) {
    // project has not a git repo => check all files
    return undefined;
  }

  const [changedFiles, stagedFiles] = await Promise.all([
    gitLsAction(argv, { mode: 'all' }),
    gitStagedAction(argv, {}),
  ]);

  const changedAndExistFiles = await existFiles(uniqueArray([...changedFiles, ...stagedFiles]), argv.root);

  return cleanAnalyseFiles(changedAndExistFiles);
};

export type { AnalyseFiles };
export { getAnalyseFiles };
