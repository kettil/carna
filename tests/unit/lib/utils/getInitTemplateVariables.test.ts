import { getInitTaskTemplateVariables } from '../../../../src/lib/utils/getInitTaskTemplateVariables';

describe('getTemplateVariables()', () => {
  test('it should work with github username', () => {
    const settings = getInitTaskTemplateVariables({ github: { name: 'username' } });

    expect(settings).toMatchSnapshot();
  });

  test('it should work without github username', () => {
    const settings = getInitTaskTemplateVariables({ github: { name: undefined } });

    expect(settings).toMatchSnapshot();
  });
});
