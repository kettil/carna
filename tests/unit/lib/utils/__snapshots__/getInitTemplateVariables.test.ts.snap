// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTemplateVariables() it should work with github username 1`] = `
Object {
  "GITHUB_USERNAME": "username",
}
`;

exports[`getTemplateVariables() it should work without github username 1`] = `
Object {
  "GITHUB_USERNAME": undefined,
}
`;
